### PR TITLE
516 system.net

### PR DESCRIPTION
**Pull request by Clockwork-Muse**
Removing custom error checking from System.Net.*
(ignoring System.Net.Sockets for the moment)

@stephentoub - I'm planning on doing a squash/overall comment like for #3020 .

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @davidsh, @cipop, @pgavlin

**Comment by davidsh** [comment]
LGTM

File: src/System.Net.Security/tests/FunctionalTests/ClientAsyncAuthenticateTest.cs [hunk 1]
```diff
@@ -62,24 +78,19 @@ public void ClientAsyncAuthenticate_EachProtocol_Success()
...
         }
 
-        [Fact]
-        public void ClientAsyncAuthenticate_MismatchProtocols_Fails()
+        [Theory]
+        [MemberData("ProtocolMismatchData")]
+        public void ClientAsyncAuthenticate_MismatchProtocols_Fails(SslProtocols server, SslProtocols client)
         {
-            foreach (SslProtocols serverProtocol in s_eachSslProtocol)
+            if ((server.ToString() == "Tls12" && new[] { "Ssl3", "Tls", "Tls11" }.Contains(client.ToString()))
```
Comment (CIPop) [inline src/System.Net.Security/tests/FunctionalTests/ClientAsyncAuthenticateTest.cs]: This looks like more code that makes the test harder to understand.

File: src/System.Net.Security/tests/FunctionalTests/ClientAsyncAuthenticateTest.cs [hunk 2]
```diff
@@ -62,24 +78,19 @@ public void ClientAsyncAuthenticate_EachProtocol_Success()
...
+                || (server.ToString() == "Tls11" && new[] { "Ssl3", "Tls" }.Contains(client.ToString()))
+                || (server.ToString() == "Tls" && client.ToString() == "Ssl3"))
```
Comment (CIPop) [inline src/System.Net.Security/tests/FunctionalTests/ClientAsyncAuthenticateTest.cs]: If this aims to be a data-driven test, then you shouldn't have the ProtocolMismatchData generate things automatically, but clearly create an array with the pairs that you want to test.

**Comment by CIPop** [comment]
/cc: @bartonjs, @vijaykota

Looks too complicated and harder to read for S.N.Security. I'd rather have a very clear list of what's getting tested. In the original implementation, I get what the code does immediately:
` if (serverProtocol != clientProtocol)`
... and you can connect then it failed.

File: src/System.Net.Security/tests/FunctionalTests/ClientAsyncAuthenticateTest.cs (hunk 2 above)
Comment (Clockwork-Muse) [inline src/System.Net.Security/tests/FunctionalTests/ClientAsyncAuthenticateTest.cs]: .... you mean add the exception to part of the generated data?

Comment (stephentoub) [inline src/System.Net.Security/tests/FunctionalTests/ClientAsyncAuthenticateTest.cs]: Yes, I'd suggest just having the MemberData yield the clientProtocol, the serverProtocol, and the expected exception Type.  There are only I believe 20 combinations to be tested.

**Comment by Clockwork-Muse** [comment]
@stephentoub - more like that?  Note that for System.Net.Security I allowed the commit of the removal of the .nugget.props file.

Also, System.Net.Security seems to be presenting strangely in VS - all the individual projects are in folders (the ones in the file system), instead of being at the root level when loaded.

**Comment by stephentoub** [comment]
> more like that?

Personally, I still find that confusing, in particular all the conditionals and ToString checks and whatnot.  I realize it's a few more lines, but I was more like this (not sure if this is what @cipop had in mind, too), just listing out each pair and the expected exception type for each:
```C#
yield return new object[] { SslProtocols.Ssl2, SslProtocols.Ssl3, typeof(IOException) };
yield return new object[] { SslProtocols.Ssl2, SslProtocols.Tls, typeof(IOException) };
yield return new object[] { SslProtocols.Ssl2, SslProtocols.Tls11, typeof(IOException) };
yield return new object[] { SslProtocols.Ssl2, SslProtocols.Tls12, typeof(IOException) };

yield return new object[] { SslProtocols.Ssl3, SslProtocols.Ssl2, typeof(IOException) };
yield return new object[] { SslProtocols.Ssl3, SslProtocols.Tls, typeof(IOException) };
yield return new object[] { SslProtocols.Ssl3, SslProtocols.Tls11, typeof(IOException) };
yield return new object[] { SslProtocols.Ssl3, SslProtocols.Tls12, typeof(IOException) };

yield return new object[] { SslProtocols.Tls, SslProtocols.Ssl2, typeof(IOException) };
yield return new object[] { SslProtocols.Tls, SslProtocols.Ssl3, typeof(AuthenticationException) };
yield return new object[] { SslProtocols.Tls, SslProtocols.Tls11, typeof(IOException) };
yield return new object[] { SslProtocols.Tls, SslProtocols.Tls12, typeof(IOException) }; 

yield return new object[] { SslProtocols.Tls11, SslProtocols.Ssl2, typeof(IOException) };
yield return new object[] { SslProtocols.Tls11, SslProtocols.Ssl3, typeof(AuthenticationException) };
yield return new object[] { SslProtocols.Tls11, SslProtocols.Tls, typeof(AuthenticationException) };
yield return new object[] { SslProtocols.Tls11, SslProtocols.Tls12, typeof(IOException) };

yield return new object[] { SslProtocols.Tls12, SslProtocols.Ssl2, typeof(IOException) };
yield return new object[] { SslProtocols.Tls12, SslProtocols.Ssl3, typeof(AuthenticationException) };
yield return new object[] { SslProtocols.Tls12, SslProtocols.Tls, typeof(AuthenticationException) };
yield return new object[] { SslProtocols.Tls12, SslProtocols.Tls11, typeof(AuthenticationException) };
```

**Comment by pgavlin** [comment]
+1 to @stephentoub's suggestion. I had in mind a simple `object[][]` property with the same contents as the iterator-method approach given.

**Comment by CIPop** [comment]
@stephentoub That's exactly what I had in mind. It's more verbose than the original but it's very clear and descriptive of the expected behavior. 

Another perspective:
I would have expected that matrix in the design spec for SslStream. Since we don't have those anymore, we can add it to the test and consider that the official design spec.

**Comment by Clockwork-Muse** [comment]
Ah, okay, give me a minute.

The reason I was resisting doing that was to try to piggyback off of any additions to the general test data, so new comparisons would be 'automatically' included.
...I'm not quite sure why I thought I needed the `ToString()`...

**Comment by Clockwork-Muse** [comment]
@stephentoub , @CIPop - done.  Thanks, Other Stephen, for the list.

**Comment by stephentoub** [comment]
LGTM, thanks.

**Comment by CIPop** [comment]
Thanks for addressing the issues and enhancing the tests @Clockwork-Muse!

**Pull request merged**